### PR TITLE
store: add ColdDatabase Database interface

### PR DESCRIPTION
**Pull request by mina86**
Introduce ColdDatabase struct which wraps a real Database interface
and transparently translates reads and writes as expected by the users
of Database interface to reads and writes as they will happen in cold
storage.  This is needed because we want to optimise cold storage
slightly (e.g. by simplifying key in State column and removing
reference counts).

### CODE REVIEW

File: core/store/src/db/colddb.rs [hunk 1]
```diff
@@ -0,0 +1,459 @@
...
+/// Some of the data we’re storing in cold storage is saved in slightly
+/// different format than it is in the main database.  Specifically:
+///
+/// - Reference counts are not saved.  Since data is never removed from cold
+///   storage, reference counts are not preserved and when fetching data are
+///   always assumed to be one.
+///
+/// - Keys of DBCol::State column are not prefixed by ShardUId.  It is not
+///   necessary to disambiguate the value and is used on hot storage to but
+///   different shards in different ranges of keys.
```
Comment (matklad) [inline core/store/src/db/colddb.rs:15]: can't parse this sentence

Comment (mina86) [reply core/store/src/db/colddb.rs:15]: s/but/put/

File: core/store/src/db/colddb.rs [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+///
+/// - Reference counts are not saved.  Since data is never removed from cold
+///   storage, reference counts are not preserved and when fetching data are
+///   always assumed to be one.
+///
+/// - Keys of DBCol::State column are not prefixed by ShardUId.  The prefix
+///   isn’t needed to disambiguate the value.  It is used on hot storage to
+///   provide some isolation by putting data of different shards in different
+///   ranges of keys.  This is not beneficial is cold storage so we’re not using
+///   the prefix to reduce space usage.
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:17]: ".. and to avoid heavy data migration in case of resharding.."

File: core/store/src/db/colddb.rs [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+/// iterations, are limited and not implemented for all columns.  When trying to
+/// implement over unsupported column the code will panic.  Refer to particular
+/// iteration method in Database trait implementation for more details.
+///
+/// Lastly, since no data is ever deleted from cold storage, trying to decrease
+/// reference of a value count or delete data is ignored and if debug assertions
+/// are enabled will cause a panic.
+struct ColdDatabase<D: Database = crate::db::RocksDB>(D);
+
+// impl<D: Database + 'static> ColdDatabase<D> {
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:39]: remove?

File: core/store/src/db/colddb.rs [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+//     pub(crate) fn new(db: D) -> std::sync::Arc<dyn Database> {
+//         std::sync::Arc::new(Self(db))
+//     }
+// }
+
+impl<D: Database> ColdDatabase<D> {
+    fn get_impl(&self, col: DBCol, key: &[u8]) -> std::io::Result<Option<Vec<u8>>> {
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:46]: comment

File: core/store/src/db/colddb.rs [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+    ///
+    /// This is implemented only for a few columns.  Specifically for Block,
+    /// BlockHeader, ChunkHashesByHeight, EpochInfo and StateDlInfos.  It’ll
+    /// panic if used for any other column.
+    ///
+    /// Furthermore, because key of ChunkHashesByHeight is modified in cold
+    /// storage, the order of iteration of that column is different than if it
+    /// would be in hot storage.
+    fn iter<'a>(&'a self, column: DBCol) -> DBIterator<'a> {
+        // Those are the only columns we’re ever iterating over.
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:83]: Do we use this iterator on all those columns?

I can understand the EpochInfo (as it is 'relatively' small), but do we have a use case for iterating over all the blocks or block headers in cold storage?

File: core/store/src/db/colddb.rs [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+        assert!(
+            matches!(
+                column,
+                DBCol::StateDlInfos
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:87]: side comment -- I actually wonder if we shoudl keep StateDlInfo in cold storage at all.

(as AFAIK this is used only when others are asking us for state sync -- which should happen only for the most recent epochs).

File: core/store/src/db/colddb.rs [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+                    | DBCol::BlockHeader
+                    | DBCol::Block
+                    | DBCol::ChunkHashesByHeight
+                    | DBCol::EpochInfo
+            ),
+            "iter on cold storage is not supported for {column}"
+        );
+        let it = self.0.iter_raw_bytes(column);
+        if column == DBCol::ChunkHashesByHeight {
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:96]: do we have a use case for iterating over this column?

File: core/store/src/db/colddb.rs [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+
+    /// Unimplemented; always panics.
+    fn iter_raw_bytes<'a>(&'a self, _column: DBCol) -> DBIterator<'a> {
+        // We’re actually never call iter_raw_bytes on cold store.
+        unreachable!();
+    }
+
+    fn write(&self, mut transaction: DBTransaction) -> std::io::Result<()> {
+        let mut idx = 0;
+        while idx < transaction.ops.len() {
```
Comment (mm-near) [inline core/store/src/db/colddb.rs:136]: comment that here we're 'filtering out' the bad operations.

**Review by mm-near — APPROVED** [review]
Nice !

File: core/store/src/db/colddb.rs (hunk 5 above)
Comment (mina86) [reply core/store/src/db/colddb.rs:83]: ChunkHashesByHeight and EpochInfo comes from state viewer.  Block and BlockHeader comes from tests.  So yes, in production we don’t iterate over those columns.  I could get rid of the implementation all together but I think keeping ChunkHashesByHeight might be good to showcase how to deal with height keys.

File: core/store/src/db/colddb.rs (hunk 6 above)
Comment (mina86) [reply core/store/src/db/colddb.rs:87]: Yes, you’re right. StadDlInfos doesn’t exist in cold storage.  I just mechanically included all columns that we ever iterate over.

File: core/store/src/db/colddb.rs (hunk 7 above)
Comment (mina86) [reply core/store/src/db/colddb.rs:96]: As per other comment, state viewer does it.

**Pull request merged**